File: leetcode/solutions/0141.LinkedListCycle/hasCycle_test.ts
import { assertEquals } from "https://deno.land/std/testing/asserts.ts";
import hasCycle from "./hasCycle.ts";
import { CreateListNode } from "../../../datastructure/SinglyLinkedListNode.ts";

Deno.test("0141. Linked List Cycle", () => {
  const linkedListNode1 = CreateListNode([3, 2, 0, 4]);
  linkedListNode1!.next!.next!.next = linkedListNode1!.next;
  assertEquals(hasCycle(linkedListNode1), true);

  const linkedListNode2 = CreateListNode([1, 2]);
  linkedListNode2!.next!.next = linkedListNode2!.next;
  assertEquals(hasCycle(linkedListNode2), true);

  const linkedListNode3 = CreateListNode([1]);
  linkedListNode3!.next = linkedListNode3;
  assertEquals(hasCycle(linkedListNode3), true);

  assertEquals(hasCycle(CreateListNode([3, 2, 0, 4])), false);
});

File: leetcode/solutions/0141.LinkedListCycle/hasCycle_test.go
package solution

import (
	"testing"

	"github.com/google/go-cmp/cmp"
	ds "github.com/yukisato/leetcode-util/datastructure"
)

func Test_0141_LinkedListCycle(t *testing.T) {
	t.Parallel()

	type Test struct {
		in  *ListNode
		out bool
	}
	var tests []Test

	node := ds.CreateListNode([]int{1, 2})
	node.Next.Next = node

	tests = append(tests, Test{node, true})

	for i, tt := range tests {
		i := i
		tt := tt

		t.Run("Test", func(t *testing.T) {
			t.Parallel()

			got := hasCycle(tt.in)

			if !cmp.Equal(got, tt.out) {
				t.Errorf("\n#%02d\ngot: %#v\nwant: %#v", i, got, tt.out)
			}
		})
	}
}
